src/inspector_profiler.cc

```diff
@@ -503,6 +507,217 @@ static void StopCoverage(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
+void ProfileWatchdog::Start() {
+  if (state() == ProfileWatchdogState::kInitialized) {
+    set_state(ProfileWatchdogState::kRunning);
+    SetTimeout();
+    int rc = uv_thread_create(&thread_, &ProfileWatchdog::Run, this);
+    CHECK_EQ(0, rc);
+  }
+}
+
+void ProfileWatchdog::Stop() {
+  if (state() != ProfileWatchdogState::kClosing &&
+      state() != ProfileWatchdogState::kClosed) {
+    set_state(ProfileWatchdogState::kClosing);
+  }
+}
+
+void ProfileWatchdog::Start(const FunctionCallbackInfo<Value>& args) {
+  ProfileWatchdog* watchdog = Unwrap<ProfileWatchdog>(args.Holder());
+  if (watchdog) {
+    watchdog->Start();
+  }
+}
+
+void ProfileWatchdog::Stop(const FunctionCallbackInfo<Value>& args) {
+  ProfileWatchdog* watchdog = Unwrap<ProfileWatchdog>(args.Holder());
+  if (watchdog) {
+    watchdog->Stop();
+  }
+}
+
+void ProfileWatchdog::Run(void* arg) {
+  ProfileWatchdog* wd = static_cast<ProfileWatchdog*>(arg);
+  uv_run(&wd->loop_, UV_RUN_DEFAULT);
+}
+
+void ProfileWatchdog::Timer(uv_timer_t* timer) {
+  ProfileWatchdog* w = ContainerOf(&ProfileWatchdog::timer_, timer);
+  Environment* env = w->env();
+  if (!env || env->is_stopping()) {
+    return;
+  }
+  if (w->state() == ProfileWatchdogState::kClosed) {
+    return;
+  }
+  env->RequestInterrupt([watchdog = std::move(w)](Environment* env) {
+    if (env->is_stopping() ||
+        watchdog->state() == ProfileWatchdogState::kClosed) {
+      return;
+    }
+    if (watchdog->state() == ProfileWatchdogState::kClosing) {
+      delete watchdog;
+      return;
+    }
+    if (watchdog->TimeoutHandler()) {
+      watchdog->AddTask(
+          [watchdog = std::move(watchdog)]() { watchdog->SetTimeout(); });
+    }
+  });
+}
+
+template <typename Fn>
+void ProfileWatchdog::AddTask(Fn&& cb, CallbackFlags::Flags flags) {
+  auto callback = tasks_.CreateCallback(std::move(cb), flags);
+  {
+    Mutex::ScopedLock lock(task_mutex_);
+    tasks_.Push(std::move(callback));
+  }
+  uv_async_send(&async_);
+}
+
+void ProfileWatchdog::HandleTasks() {
+  while (tasks_.size() > 0) {
```

Comment:
You need to be holding the mutex before checking the size. Also:
```suggestion
  while (!tasks_empty()) {
```

Reply:
Why? I'm assuming this code path will be only executed by the thread created on `::Start()`

Reply:
Oh, I take that back. The comment for that method says `size() is atomic and may be called from any thread`.

I'm looking at `src/callback_queue-inl.h` right now and the way it uses sequentially consistent atomic loads and stores for `size_` isn't very optimal but I guess it's correct (if not obviously so.)